Generate configuration schema from C and drop Perl and trang dependencies

k0ekk0ek opened the pull request:
@eboasson, in PR #441 building documentation was added. Thereby "introducing" dependencies on `trang` (and with that Java) and `perl`. I think everybody involved agrees that droping dependencies on said packages is a good thing, especially because both are only required for something as trivial as generating an `.rnc` and `.xsd`. I thought about it a little and came up the changes in this PR. It's not finished and this is merely a quick request for comments before I continue with the changes.

Basically I isolated all configuration options in a separate file so that the options can be iterated without needing all the includes in `q_config.c`, I used macros instead of an actual `struct cfgelem` so that I elements can discarded, e.g. the init, print, etc functions. I would've gone with a function in `q_config.c`, but the options need to be recompiled because the `BLURB` isn't included otherwise. The `ddsconf` tool would be used to generate `.xsd` and `.rnc` files instead.

Please let me know if this stands a chance of being included to replace `makernc.pl` and `trang`. Thanks.

eboasson commented:
Hi @k0ekk0ek there are some things I am less than happy with but some creative use of the C preprocessor usually gets a smile from me 😁 especially if it isn't too tricky. I definitely like the why-did-I-not-think-of-that obviousness of using a C program that contains the tables instead of a fragile parser like in "makernc.pl".

For the rest, well, I think it needs another header file to define the DSL that is formed by the macros from `NODATA` to `MOVED` and the type definition for the configuration elements. That way, their being shared between the three places is a bit more clear. I'd probably put
```C
#ifndef BLURB
#define BLURB(text) text
#endif 
```
in there and then do `#define BLURB(text) NULL` in `q_config.c`.

But overall, sure. It pretty trivially eliminates the perl-y bits and once you figure out the rules of XSD we are in a much better position. It won't win a beauty contest, but that's already better than the current code, which is sure to lose it. So that can hardly be an argument against going down this path.

k0ekk0ek commented:
Thanks for taking a look @eboasson. I agree there's some rough edges. I'll take care of those, I have some ideas, but didn't want to do too much work prior to you having a look, exactly because it's a little hacky :wink:.

rotu commented:
> in PR #441 building documentation was added. Thereby "introducing" dependencies on `trang` (and with that Java) and `perl`.

These dependencies existed beforehand - I merely made the build process easier for them. You can call me names, but I will not be called a Perl nor a Java developer. Python and C++ forever!

k0ekk0ek commented:
You're absolutely right @rotu. "introducing" was quoted to acknowledge that, but maybe I should've phrased it a little better.

eboasson commented:
At least I interpreted it in the way @k0ekk0ek meant it, and anyone looking at the repository would know that it was I who used them ... But @rotu, Python and C++, really? Not Haskell or Ada? 😁

rotu commented:
I quite like Haskell, and haven't tried out Ada. But APL appeals to me especially: if nobody can understand your character set, then nobody can criticize your code!

---

Anyway, as far as this PR, it's very clever and quite cool! Structural recursion is so much more natural than trees than iteration.

Another possible approach is to make this whole thing a higher order function. That is you are injecting macros lexically:
```
#define BLURB ....
#define CFGELEMS ...
...
#include "q_configopts.h"
```

And you could inject function pointers:
```
typedef void (*blurb_fn)(void * context, const struct blurb *);
typedef void (*cfgelems_fn)(void * context, const struct cfgelems *);
void for_each_config_opt(void * context, blurb_fn, cfgelems_fn, ...);
```

I don't actually know if it's any better, but I do quite like how it specifies what each handler has available to work with.

k0ekk0ek commented:
Currently I'm leaning towards making it a "language" of it's own, sort-of. Like this:
```
ELEMENTS(something_cfgelems,
  STRING("Name", <something_cfgattrs-or-NULL>, ...
    MEMBER(struct-member),
    FUNCTIONS(init, update, free, print),
    DESCRIPTION("<p>description</p>"), // q_config.c drops everything from here on
    RANGE("if;applicable"), // converted to .range = "if-applicable"
    UNIT("if-applicable")), // converted to .unit = "if-applicable"
  ENUM("...", ...
    MEMBER(...),
    FUNCTION(...),
    DESCRIPTION("...")
    VALUES("all;enum;values")), // converted to .values = "all;enum;values"
  BOOL(...),
  INT(...),
  LIST(..., VALUES("all;possible;list;values")) // the type formerly known as Comma
);
```

That way all information needed for both `q_config.c` and `ddsconf` are kept in one place, each only takes what it needs and there's no need to reuse the original structs. `q_configopts.h` really only will contain the above.

All-in-all it's pretty self-explanatory and quite readable. Maybe it could even win a beauty contest(?) :wink:

k0ekk0ek commented:
So... I learned a lot more about xsd than I intended to :grin: This now contains something that generates `.xsd` and `.rnc` files straight from C so we can drop the Perl and trang dependencies. I need to spit shine it a little bit, but wanted to check if you agree with the proposed changes @eboasson. Mostly referring to macro's, I wouldn't bother reviewing everything else yet.

k0ekk0ek commented:
`ddsconf` can generate Markdown now too. Imho it the units must be moved from `q_configopts.h` to something like `ddsi_units.h` or something (moved them from `makernc.pl`) and rename `q_configopts.h` itself to `ddsi_options.h`. Apart from that there's a couple of other places that need some polishing, but the mechanism itself can be reviewed.

Note that:
I'm planning on replacing `printwrap` by something that wraps lines when sanitizing the description, and not all `cfgelem->meta` members are freed, but that's something I'll do after the initial review.

eboasson commented:
I do think this is definitely an improvement over the current situation. I do need to have a better look at the code and what it generates, but I'm leaning towards taking this and removing the old junk.

k0ekk0ek commented:
Actually working on polishing this a bit more. There's some things that can be simplified without too much effort. @eboasson, any reason we want to stick to the 80 column limit? Dropping that makes things easier and since we use the `<p>` tags and such for formatting, the end result should be the same in browsers etc(?)

eboasson commented:
> Actually working on polishing this a bit more. There's some things that can be simplified without too much effort. @eboasson, any reason we want to stick to the 80 column limit?

I seem to recall it was either the documentation pop-up in VSCode or something Emacs-related that made me format it like that. I don't really think the Emacs bit is relevant, but the VSCode bit is worth checking.

> Dropping that makes things easier and since we use the `<p>` tags and such for formatting, the end result should be the same in browsers etc(?)

Now it goes from HTML-ish to markdown; it might be more sensible to go from markdown to HTML-ish. But don't waste too much time on it. If functionally it does all we want and the only problem is the aesthetics of the documentation pop-up, I think we're already in a pretty decent shape.

k0ekk0ek commented:
@eboasson, Windows is causing some problems still and I'll have to update `.travis.yml` to test whether the files differ from the ones in the repository, but I think this is ready for review. `diff`ing the existing files with the new output shows some errors in the existing `.xsd` and `.rnc` files and mainly updated comments. I left the line-wrap in for now, we can remove it in future. Main reason for removing it would be that it handles a Markdown special (does line start with ` *`) that I don't particularly like.

Nice addition for the future maybe to include `.rst` so that we can point to individual configuration options from the manual etc.

I'll handle the Windows errors one of these days (have to setup a Windows machine first).

k0ekk0ek commented:
A typicall case of `__VA_ARGH!__`

If you thought you've seen it all, given a snippet like:
```
ELEMENTS(foo_cfgelems,
  GROUP("bar" ...)
#if DDSI_INCLUDE_BAZ
  GROUP("baz" ...)
#endif
);
```

Visual Studio will throw an error like:
```
Z:\path\to\source\file.c(<line>,1): error C2121: '#': invalid character: possibly the result of a macro expansion [Z:\path\to\build\project.vcxproj]
```

Even if the macro is defined...

eboasson commented:
I wouldn't be surprised if Microsoft got that one right (as in: following the spec), unlikely though it may seem. I suppose the most satisfying solution is to dump Visual Studio altogether, but the more pragmatic one would be to do:

```C
#ifdef DDSI_INCLUDE_BAZ
#define GROUP_BAZ GROUP("baz" ...) ,
#else 
#define GROUP_BAZ
#endif
```

although that comma is quite ugly and messes-up auto-indenters ... expanding to a dummy entry if `DDSI_INCLUDE_BAZ` isn't defined could make sense.

But I'm sure I'm not suggesting anything you hadn't thought of already.

reicheratwork commented on src/core/ddsi/include/dds/ddsi/ddsi_cfgelems.h:
```diff
@@ -0,0 +1,1994 @@
+      "<p>This element controls the interval at which to check whether "
+      "threads have been making progress.</p>"),
+    UNIT("duration"),
+    RANGE("100ms;1hr")),
```
Was this present in q_config.c?

k0ekk0ek replied:
I thought I'd save myself the hassle of looking up the ranges in a separate table as done in `makernc.pl` and specify it as a field instead (`ddsconf` only, not for `ddsi`). The function `uf_duration_100ms_1hr` gets translated by the script to `100ms;1hr` (as the name correctly suggests :smile:).

reicheratwork commented on src/core/ddsi/include/dds/ddsi/ddsi_cfgelems.h:
```diff
@@ -0,0 +1,1994 @@
+      "has been multicasted too recently. This setting has no effect on "
+      "unicasted retransmits.</p>\n"
+      "<p>See also Internal/RetransmitMerging.</p>"),
+    UNIT("duration"),
```
This is also something that was not present in the q_config.c fields.?

k0ekk0ek replied:
See other comment about units etc.

reicheratwork commented on src/core/ddsi/include/dds/ddsi/ddsi_cfgelems.h:
```diff
@@ -0,0 +1,1994 @@
+      "always try to merge.</li></ul>\n"
+      "<p>The default is <i>never</i>. See also "
+      "Internal/RetransmitMergingPeriod.</p>"),
+    VALUES("never","adaptive","always")),
```
The values field is something new extending the q_config.c definition?

k0ekk0ek replied:
Same

reicheratwork commented on src/core/ddsi/include/dds/ddsi/ddsi_cfgelems.h:
```diff
@@ -0,0 +1,1994 @@
+};
+
+static struct cfgelem root_cfgattrs[] = {
+  NOP("xmlns"),
```
Missing default value ""?

k0ekk0ek replied:
NOP isn't really used etc.

k0ekk0ek commented:
Removed line wrapping after all. Visual Studio didn't complain about the generated rnc and xsd files.

Renamed `DDSSecurity` to `Security|DDSSecurity` as that fits a little better with the rest.

Updated xsd generation a bit with regards to `sequence`, `all` and `choice` elements in `complexType` elements and made `minOccurs` and `maxOccurs` output like with `trang`. When generating xsd from the generated rnc, the documents are identical. Well, almost. `trang` contains a bug (I think) that shows with the xsd for the Domain element. It'll use choice because the Threads element has `maxOccurs="unbounded"`, but `all` must be used as the `Threads` element may only occur once. It's the `Threads` element child `Thread` that's unbounded. At first I thought this was caused by the output from `makernc.pl` because it got the multiplicity for a number of groups wrong (used `*` instead of `?`, see below), but it showed with the newly generated output as well. Funnily enough, `trang` got the xsd output mostly right even with the incorrect rnc.

* Compatibility
* Discovery
* Internal
* IgnoredPartitions
* NetworkPartitions
* PartitionMapping
* PartitionMappings
* SSL
* Sizing
* TCP
* ThreadPool
* Threads
* Tracing

It builds on all platforms, schema files and documentation were updated. I think this is a wrap. I can see why the original author only generated rnc, let's hope we won't have to write xsd generating code again any time soon :sweat_smile: :wink:


EDIT:
 * Reverted Domain from unbounded back to max occurrence of 1
 * Reverted Group from unbounded back to max occurrence of 1
 * Reverted Security from max occurrence of to unbounded

Group maybe correct(?), but I'm not sure about the other two. Seems multiple Domain elements should be allowed (multi-domain)? Can Security be configured more than once?

eboasson commented on .travis.yml:
```diff
@@ -218,11 +217,15 @@ script:
   - |
-    if [ ${BUILD_SCHEMA} ]; then
-      git diff --exit-code; # check that no files (e.g. generated schemas) are changed
+    if [ "${SSL}" = "YES" ] && [ "${SECURITY}" = "YES" ]; then
```
Methinks this is wrong. Besides, I think we should just remove `BUILD_SCHEMA` altogether. It takes no time, so why make it optional?

k0ekk0ek replied:
The schema is different if SSL and/or security is not included so the check would fail. I'm afraid there's no way around that...

eboasson replied:
Why did I not think of that!

eboasson commented on docs/CMakeLists.txt:
```diff
@@ -8,36 +8,28 @@
+    COMMAND ddsconf ARGS -f rnc -o "${cyclonedds_rnc}"
+    COMMAND ddsconf ARGS -f xsd -o "${cyclonedds_xsd}"
+    DEPENDS ddsconf)
+  add_custom_target(schema DEPENDS "${cyclonedds_rnc}" "${cyclonedds_xsd}")
```
Just like I think `BUILD_SCHEMA` is now something to be removed, I think these should be generated always.

k0ekk0ek replied:
Done. schemas are always generated except when cross compiling.

eboasson commented on etc/cyclonedds.xsd:
```diff
@@ -17,18 +17,18 @@ CycloneDDS configuration</xs:documentation>
         <xs:element ref="config:Internal"/>
         <xs:element ref="config:Partitioning"/>
         <xs:element ref="config:SSL"/>
+        <xs:element maxOccurs="unbounded" ref="config:Security"/>
```
This is weird: why is does this one end up as "unbounded" and the rest doesn't?

k0ekk0ek replied:
maxOccurs has a default of 1, which seems about right, even if the parser itself doesn't care about it?

eboasson replied:
My main issue is that it is weird that it is different in this one case. Any "normal" configuration I have seen has each section listed at most once, so both maxOccurs="1" and maxOccurs="unbounded" would work as far as I am concerned.

k0ekk0ek replied:
Fixed

eboasson commented on etc/cyclonedds.xsd:
```diff
@@ -44,17 +44,16 @@ CycloneDDS configuration</xs:documentation>
           <xs:documentation>
-&lt;p&gt;Domain id this configuration applies to, or "any" if it applies to all
-domain ids.&lt;/p&gt;&lt;p&gt;The default value is: &amp;quot;any&amp;quot;.&lt;/p&gt;</xs:documentation>
+&lt;p&gt;Domain id this configuration applies to, or "any" if it applies to all domain ids.&lt;/p&gt;
```
Shouldn't the " characters be turned into &quot;? (Maybe no-one cares, and certainly the original &amp;quot; was broken.)

k0ekk0ek replied:
I translated all xml entities at first, but `trang` didn't replace the ones in the comment, so I figured it'd be more readable like this. So it isn't required, but I can reinstate if desired.

eboasson replied:
If a plain double-quote character works, it is good enough for me :)

eboasson commented on etc/cyclonedds.xsd:
```diff
@@ -17,18 +17,18 @@ CycloneDDS configuration</xs:documentation>
     </xs:annotation>
     <xs:complexType>
-      <xs:choice minOccurs="0" maxOccurs="unbounded">
+      <xs:choice minOccurs="0">
```
This seems to not be quite right: VS Code with the RedHat XML extensions (the version I have anyway) complain that a file with more than one element under Domain is invalid, and it seems to match the code here. I tried hacking it:
```diff
diff --git a/etc/cyclonedds.xsd b/etc/cyclonedds.xsd
index d38f2ea4..99558180 100644
--- a/etc/cyclonedds.xsd
+++ b/etc/cyclonedds.xsd
@@ -17,17 +17,17 @@ CycloneDDS configuration</xs:documentation>
 &lt;p&gt;The General element specifying Domain related settings.&lt;/p&gt;</xs:documentation>
     </xs:annotation>
     <xs:complexType>
-      <xs:choice minOccurs="0">
-        <xs:element ref="config:Compatibility"/>
-        <xs:element ref="config:Discovery"/>
-        <xs:element ref="config:General"/>
-        <xs:element ref="config:Internal"/>
-        <xs:element ref="config:Partitioning"/>
-        <xs:element ref="config:SSL"/>
-        <xs:element maxOccurs="unbounded" ref="config:Security"/>
-        <xs:element ref="config:Sizing"/>
-        <xs:element ref="config:TCP"/>
-        <xs:element ref="config:ThreadPool"/>
+      <xs:all>
+        <xs:element minOccurs="0" ref="config:Compatibility"/>
+        <xs:element minOccurs="0" ref="config:Discovery"/>
+        <xs:element minOccurs="0" ref="config:General"/>
+        <xs:element minOccurs="0" ref="config:Internal"/>
+        <xs:element minOccurs="0" ref="config:Partitioning"/>
+        <xs:element minOccurs="0" ref="config:SSL"/>
+        <xs:element minOccurs="0" ref="config:Security"/>
+        <xs:element minOccurs="0" ref="config:Sizing"/>
+        <xs:element minOccurs="0" ref="config:TCP"/>
+        <xs:element minOccurs="0" ref="config:ThreadPool"/>
         <xs:element minOccurs="0" name="Threads">
           <xs:annotation>
             <xs:documentation>
@@ -40,7 +40,7 @@ CycloneDDS configuration</xs:documentation>
           </xs:complexType>
         </xs:element>
         <xs:element ref="config:Tracing"/>
-      </xs:choice>
+      </xs:all>
       <xs:attribute name="Id">
         <xs:annotation>
           <xs:documentation>
```
and then this problem goes away. So I the use of `any`s needs to be a bit more liberal.

k0ekk0ek replied:
Strictly speaking, according to the multiplicity, the `Security` group can occur "unbounded" and `all` only allows elements that do not occur or only occur once. I'll use the xsd spec and correct it. If this is correct, does that mean that the `Security` group may appear just once?

eboasson replied:
I can live with it whichever way it goes, as long as all reasonable configuration files are valid according to the XSD. What I noticed is that one of those "reasonable" ones triggered a warning.

k0ekk0ek commented:
I'll have another look. I know a thing or about xsd and the generation is reasonably straightforward, it's the `all` vs. `choice` vs. `sequence` that's somewhat "difficult". I used the w3schools documentation and the descriptions are a bit ambiguous.

k0ekk0ek commented:
Should be good now. I've introduced the possibility for adding `MAXIMUM` (and `MINIMUM`) next to `DESCRIPTION`, `UNIT`, etc and the corresponding value is used if specified. Comments are added in `ddsi_cfgelems.h` to denote the edge cases (so contributors don't needlessly add them). I've reintroduced `trang` in `.travis.yml` to test if the generated `.xsd` matches the `.xsd` generated from the generated `.rnc` (they currently do :grin:). I've loaded up the xsd in Visual Studio Code and it seemed to work just fine. Should this PR be accepted, I propose I clean up the commits a bit and force-push so the history is nice and clean, although, on the other hand it would be good for statistics to leave it as is :wink:

k0ekk0ek commented:
Rebased & force-pushed. If the checks pass, I'll hit merge
